[R3C-465] MCH: pedestals calibration workflow

aferrero2707 opened the pull request:
This PR introduces the workflows needed to process pedestal data, build the list of bad/noisy channels and store the list in the CCDB. The bad/noisy channels list will be used by the DCS to mute the corresponding channels during front-end configuration, and by the reconstruction in order to know which channels are expected to have no signals.

The processing workflow, in a simplified form, looks like this:
```
o2-mch-pedestals-decoding-workflow -b --patch-rdh | o2-calibration-mch-channel-calib-workflow -b --pedestal-threshold 200 --noise-threshold 2.0 | o2-calibration-ccdb-populator-workflow -b
```

In reality the decoding workflows should run in parallel on various EPNs, while the calibration workflow should aggregate the information from the various decoders and compute the channel-by-channel mean and RMS of the pedestals.
Once the end-of-stream is reached the calibrator scans the pedestals and compares the values with the user-supplied thresholds. Channels exceeding either threshold are added to the bad/noisy list, which is finally inserted into the CCDB.

The CDDB data consists of a vector of 32-bit integers, each integer representing the unique ID of a readout channel.

shahor02 commented on Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h:
```diff
@@ -0,0 +1,132 @@
+  using Slot = o2::calibration::TimeSlot<o2::mch::calibration::MCHChannelData>;
+
+ public:
+  MCHChannelData()
```
I think CI will require to redefine this to ``MCHChannelData() = default``

aferrero2707 replied:
Changed.

shahor02 commented on Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibratorAlt.h:
```diff
@@ -0,0 +1,106 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+#ifndef MCH_CHANNEL_CALIBRATOR_H_
+#define MCH_CHANNEL_CALIBRATOR_H_
```
this class seem not to be used/compiled, but it has the same guards as https://github.com/AliceO2Group/AliceO2/blob/445d4cff6ca5f0385beff377b7535916a8c41c42/Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h#L11-L12, should be changed

aferrero2707 replied:
This was a test of the alternative approach I proposed at the beginning, and the files were committed by mistake. I will remove them.

shahor02 commented on Detectors/MUON/MCH/Calibration/src/MCHChannelCalibratorAlt.cxx:
```diff
@@ -0,0 +1,131 @@
+
+      mHisto[hIndex](s);
+
+      /*mEntries[solarId][dsId][channel] += 1;
```
if not needed, better remoce the commented code

aferrero2707 replied:
See above

shahor02 commented on Detectors/MUON/MCH/Calibration/include/MCHCalibration/MCHChannelCalibrator.h:
```diff
@@ -0,0 +1,132 @@
+  ClassDefNV(SampaChannelId, 1); // class for MCH readout channel
+};
+
+class BadChannelsVector
```
if your calibration object should contain nothing but the content of this vector, then it is indeed better to define ``SampaChannelId`` and ``BadChannelsVector`` in a single header in DataFormats, and use the latter as CCDB object.

aferrero2707 replied:
Done.

shahor02 reviewed (commented):
Hi @aferrero2707 
To me it looks fine, see a few minor comments below.

aferrero2707 committed:
[MCH] introduced workflow for pedestals analysis

This workflow processes pedestal data, builds the list
of bad/noisy channels and stores the list in the CCDB.

aferrero2707 commented:
Please do not merge this PR yet when all CI jobs will be green. I am still working on the interface with the QC.

aphecetche commented on DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h:
```diff
@@ -0,0 +1,63 @@
+  ClassDefNV(DsChannelId, 1); // class for MCH readout channel
+};
+
+class DsChannelGroup
```
Why do we need a ChannelGroup class which is nothing more than a ChannelId vector ?

aferrero2707 replied:
In order to transform the vector into a root object. If there is an easier and more direct way to do that, I'm open to suggestions.

aphecetche commented on DataFormats/Detectors/MUON/MCH/include/DataFormatsMCH/DsChannelGroup.h:
```diff
@@ -0,0 +1,63 @@
+namespace mch
+{
+
+class DsChannelId
```
that might not have to be solved for this PR, but ... we actually have two ways to reference the channels : (solarId,dsId,channel) (raw/online data oriented) and (deId,dsId,channel) (digit/offline oriented) and both can be encoded to 32 bits. 

I was wondering it we could not decide to call those channelIds (== uint32_t) and use a couple of bits in those 32 to indicate which convention is used (solar or de) ?

aferrero2707 replied:
Indeed we could use the most significant bit to indicate the type of encoding, and two different constructors (`DsChannelId(uint16_t solarId, uint8_t dsId, uint8_t channel)` or `DsChannelId(uint16_t deId, uint16_t padId)`) for the two cases.

As a starting point, I will add getter functions for the solarId/dsId/channel values, to avoid the direct manipulation of the 32-bit encoded value.

aferrero2707 committed:
[MCH] added decoder error processing

aferrero2707 committed:
[MCH] added output for sending the pedestal values to the QC

A "MCH/PEDESTALS" output spec has been added, which
allows to send the computed pedestal mean/rms values
to a downstream DPL device, for example to the QC.
The QC in this case can simply plot and checks the
pre-computed pedestals, without having to re-process
the raw data.

aferrero2707 commented:
@TimoWilken @ktf the fullCI check fails although there seem not to be any build errors, only warnings... any idea?

aferrero2707 commented:
@aphecetche @shahor02 for me the code in this PR is complete and ready to be merged once the CI jobs turn al green... I know that parts of the code are still perfectible, but I would propose to get this out for MW3 and then further improve the code structure once the workflow is validated with the tests.

There is a QC PR that depends on this one, and which is also needed for MW3: https://github.com/AliceO2Group/QualityControl/pull/632

TimoWilken commented:
@aferrero2707 (Cc: @ktf) that's a bug in the o2checkcode.sh build script in alidist. I've proposed a patch there to fix it, and I'll restart the test once the fix is merged.

aferrero2707 commented:
@TimoWilken thanks!

aferrero2707 commented:
@TimoWilken @ktf the o2-dataflow CI job completed again with a failure, although there are no errors in the log... if I understood correctly, this should have been fixed by alisw/alidist#2891, right?

TimoWilken commented:
@aferrero2707 that PR only fixed the fullCI check; the o2-dataflow check has a different problem, which I'm currently in the process of fixing.

aferrero2707 committed:
[MCH] added bechmarking info to pedestals processor

aferrero2707 committed:
[MCH] fixed clang formatting

aferrero2707 committed:
[MCH] fixed deinition of default constructor

aphecetche commented on Detectors/MUON/MCH/Calibration/src/pedestals-decoding-workflow.cxx:
```diff
@@ -145,7 +146,11 @@ class PedestalsTask
-
+    auto stop = [this]() {
+      LOG(INFO) << "time spent for decoding (ms): min=" << mTimeDecoderMin->count() << ", max="
+          << mTimeDecoderMax->count() << ", mean=" << mTimeDecoder.count() / mTFcount;
```
do I read correctly that this will break when running with mDebug=false as mTFcount will be zero then ?

aferrero2707 replied:
Good catch! I will put the mTFcout increment outside of the mDebug block.

aferrero2707 committed:
[MCH] avoid possible division by zero

aferrero2707 commented:
@TimoWilken the o2-dataflow job failed again, with no logs (`No logs found. Please check the aurora log.` message when looking for full build log).

aferrero2707 commented:
@aphecetche do you think we can approve this PR and merge once the CI jobs are all green?

TimoWilken commented:
@aferrero2707 thanks for the report; it was missing a tag. I've fetched it and restarted the check.
